### PR TITLE
Add secret detection logic to Azure service principal crawler

### PR DESCRIPTION
**Pull request by nkvuong**
## Changes
- Add client secret detection logic to Azure service principal crawler, this is needed for #874

### Tests

- [x] manually tested
- [x] added unit tests
- [x] added integration tests

### CODE REVIEW

File: src/databricks/labs/ucx/assessment/azure.py [hunk 1]
```diff
@@ -8,17 +8,14 @@
...
-    STORAGE_ACCOUNT_EXTRACT_PATTERN,
-    azure_sp_conf_present_check,
-    logger,
-)
+from databricks.labs.ucx.assessment.crawlers import azure_sp_conf_present_check, logger
 from databricks.labs.ucx.assessment.jobs import JobsMixin
 from databricks.labs.ucx.framework.crawlers import CrawlerBase, SqlBackend
 
+SECRET_PATTERN = r"{{secrets\/(.*)\/(.*)}}"
+TENANT_PATTERN = r"https:\/\/login.microsoftonline.com\/(.*)\/oauth2\/token"
```
Comment (nfx) [inline src/databricks/labs/ucx/assessment/azure.py:16]: Technically, this only works for public cloud, but not for govcloud or china

File: src/databricks/labs/ucx/assessment/azure.py [hunk 2]
```diff
@@ -129,33 +126,53 @@ def _safe_get_cluster_policy(self, policy_id: str) -> Policy | None:
...
+                secret_scope, secret_key = None, None
             else:
-                tenant_key = "fs.azure.account.oauth2.client.endpoint"
+                client_secret = self._get_key_from_config(matching_secret_keys[0], config, False)
+                if client_secret is not None:
+                    secret_matched = re.findall(SECRET_PATTERN, client_secret)
+                    if len(secret_matched) == 0:
+                        logger.warning('Secret in config stored in plaintext.')
+                    else:
+                        secret_scope, secret_key = secret_matched[0][0], secret_matched[0][1]
```
Comment (nfx) [inline src/databricks/labs/ucx/assessment/azure.py:162]: Nesting gets too deep, split this method up

Comment (nkvuong) [reply src/databricks/labs/ucx/assessment/azure.py]: done

File: src/databricks/labs/ucx/assessment/azure.py (hunk 1 above)
Comment (nkvuong) [reply src/databricks/labs/ucx/assessment/azure.py]: expanded regex for govcloud & cn

File: src/databricks/labs/ucx/assessment/azure.py [hunk 3]
```diff
@@ -129,33 +126,40 @@ def _safe_get_cluster_policy(self, policy_id: str) -> Policy | None:
...
+            tenant_key = "fs.azure.account.oauth2.client.endpoint"
+            client_secret_key = "fs.azure.account.oauth2.client.secret"
+
             # retrieve storage account configured with this spn
             storage_account_matched = re.search(STORAGE_ACCOUNT_EXTRACT_PATTERN, spn_app_id_key)
             if storage_account_matched:
                 storage_account = storage_account_matched.group(1).strip(".")
-                tenant_key = "fs.azure.account.oauth2.client.endpoint." + storage_account
+                # adjust the key to lookup for tenant id & secret id
+                tenant_key = tenant_key + f".{storage_account}"
```
Comment (nfx) [inline src/databricks/labs/ucx/assessment/azure.py:148]:
```suggestion
                tenant_key =
f"{tenant_key}.{storage_account}"
```
More readable this way

File: src/databricks/labs/ucx/assessment/azure.py [hunk 4]
```diff
@@ -129,33 +126,40 @@ def _safe_get_cluster_policy(self, policy_id: str) -> Policy | None:
...
+                client_secret_key = client_secret_key + f".{storage_account}"
+
+            # retrieve client secret of spn
+            matching_secret_keys = [key for key in config.keys() if re.search(client_secret_key, key)]
```
Comment (nfx) [inline src/databricks/labs/ucx/assessment/azure.py:152]: Do you need suffix matching or full regex? Regex may yield dot-related bugs

File: src/databricks/labs/ucx/assessment/azure.py [hunk 5]
```diff
@@ -8,17 +8,14 @@
...
-    SECRET_PATTERN,
-    STORAGE_ACCOUNT_EXTRACT_PATTERN,
-    azure_sp_conf_present_check,
-    logger,
-)
+from databricks.labs.ucx.assessment.crawlers import azure_sp_conf_present_check, logger
 from databricks.labs.ucx.assessment.jobs import JobsMixin
 from databricks.labs.ucx.framework.crawlers import CrawlerBase, SqlBackend
 
+SECRET_PATTERN = r"{{secrets\/(.*)\/(.*)}}"
```
Comment (nfx) [inline src/databricks/labs/ucx/assessment/azure.py:15]: Can you make these constants as private fields of a mixin class of they are not used elsewhere?

**Review by nfx — CHANGES_REQUESTED** [review]

**Pull request merged**